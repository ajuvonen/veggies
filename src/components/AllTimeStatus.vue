<script lang="ts" setup>
import {computed} from 'vue';
import {storeToRefs} from 'pinia';
import {useActivityStore} from '@/stores/activityStore';
import StatContainer from '@/components/StatContainer.vue';

const activitysStore = useActivityStore();
const {over30Veggies, atMostVeggies, getWeekStarts, uniqueVeggies, completedChallenges} =
  storeToRefs(activitysStore);

const sections = computed(() => ({
  totalWeeks: getWeekStarts.value.length,
  over30Veggies: over30Veggies.value,
  uniqueVeggies: uniqueVeggies.value.length,
  atMostVeggies: atMostVeggies.value,
  completedChallenges: completedChallenges.value,
}));
</script>
<template>
  <div class="all-time-status__container">
    <div
      v-for="([key, value], index) in Object.entries(sections)"
      :key="key"
      :class="{'col-span-2': index === 4}"
      :data-test-id="`all-time-status-${key}`"
      class="all-time-status__item"
    >
      <StatContainer :statAmount="value" :statKey="key" />
    </div>
  </div>
</template>
<style lang="scss" scoped>
.all-time-status__container {
  @apply grid grid-cols-2 grid-rows-3;
  @apply text-center;
  row-gap: 1rem;
}

.all-time-status__item {
  @apply label-like relative;
  @apply flex flex-col items-center justify-center;
}
</style>
