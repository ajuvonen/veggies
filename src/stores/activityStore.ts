import {computed, ref} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {debounceFilter, useIntervalFn, useNow, useStorage} from '@vueuse/core';
import {DateTime} from 'luxon';
import {
  countBy,
  difference,
  entries,
  intersection,
  map,
  pipe,
  prop,
  sortBy,
  take,
  takeWhile,
} from 'remeda';
import {
  Category,
  type Favorites,
  type Challenge,
  type Week,
  type Achievements,
  AchievementLevel,
} from '@/utils/types';
import {
  achievementLevelHelper,
  dateParser,
  dateReplacer,
  getCategoryForVeggie,
  getRandomVeggie,
} from '@/utils/helpers';
import {useAppStateStore} from '@/stores/appStateStore';
import {
  BEANS,
  BOTANICAL_BERRIES,
  FRUITS,
  GRAINS,
  LEAFIES,
  MUSHROOMS,
  NUTS,
  RED_VEGGIES,
  ROOTS,
  VEGETABLES,
} from '@/utils/constants';

export const useActivityStore = defineStore('activity', () => {
  const {settings} = storeToRefs(useAppStateStore());
  const currentDate = ref(DateTime.now());
  useIntervalFn(() => {
    const now = DateTime.now();
    if (!currentDate.value.hasSame(now, 'day')) {
      currentDate.value = now;
    }
  }, 2000);

  // State refs
  const startDate = useStorage<DateTime | null>('veggies-start-date', null, localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? DateTime.fromISO(v.split('T')[0]) : null),
      write: (v) => v?.toISODate() ?? '',
    },
  });

  const weeks = useStorage<Week[]>('veggies-weeks', [], localStorage, {
    mergeDefaults: true,
    eventFilter: debounceFilter(2000),
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  const challenges = useStorage<Challenge[]>('veggies-challenges', [], localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  // Computed getters
  const currentWeekStart = computed(() => currentDate.value.startOf('week'));

  const getWeekStarts = computed(() => {
    if (!startDate.value) return [currentWeekStart.value];
    const totalWeeks = Math.ceil(currentDate.value.diff(startDate.value, 'week').weeks);
    return Array.from({length: totalWeeks}, (_, weekIndex) =>
      currentWeekStart.value.minus({weeks: weekIndex}),
    );
  });

  const hotStreak = computed(() => {
    let over30Weeks = takeWhile(
      getWeekStarts.value,
      (weekStart, weekIndex) => weekIndex === 0 || veggiesForWeek.value(weekStart).length >= 30,
    ).length;
    if (currentVeggies.value.length < 30) {
      over30Weeks--;
    }
    return over30Weeks;
  });

  const allVeggies = computed(() => weeks.value.flatMap(prop('veggies')));

  const uniqueVeggies = computed(() => [...new Set(allVeggies.value)]);

  const over30Veggies = computed(
    () => weeks.value.filter(({veggies}) => veggies.length >= 30).length,
  );

  const atMostVeggies = computed(() =>
    Math.max(...weeks.value.map(({veggies}) => veggies.length), 0),
  );

  const completedChallenges = computed(
    () =>
      challenges.value.filter(({startDate, veggie}) =>
        veggiesForWeek.value(startDate).includes(veggie),
      ).length,
  );

  const veggiesForWeek = computed(
    () => (weekStart: DateTime) =>
      weeks.value.find(({startDate}) => startDate.equals(weekStart))?.veggies ?? [],
  );

  const currentVeggies = computed({
    get: () => veggiesForWeek.value(currentWeekStart.value),
    set: (veggies: string[]) => setVeggiesForWeek(veggies),
  });

  const currentChallenge = computed(
    () => challenges.value.find(({startDate}) => startDate.equals(currentWeekStart.value))?.veggie,
  );

  const suggestions = computed(() =>
    pipe(
      allVeggies.value.filter((veggie) => !currentVeggies.value.includes(veggie)),
      countBy((veggie) => veggie),
      entries(),
      sortBy([prop(1), 'desc']),
      take(settings.value.suggestionCount),
      map(prop(0)),
    ),
  );

  const favorites = computed(() =>
    Object.values(Category).reduce(
      (acc, category) => ({
        ...acc,
        [category]: pipe(
          allVeggies.value.filter((veggie) => getCategoryForVeggie(veggie) === category),
          countBy((veggie) => veggie),
          entries(),
          sortBy([prop(1), 'desc']),
          take(6),
        ),
      }),
      {} as Favorites,
    ),
  );

  const weeklyAchievements = computed(() => (veggies: string[] = currentVeggies.value) => ({
    allOnRed: achievementLevelHelper(
      [[10, AchievementLevel.Gold]],
      intersection(RED_VEGGIES, veggies).length,
    ),
    botanicalBerries: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BOTANICAL_BERRIES, veggies).length,
    ),
    goNuts: achievementLevelHelper(
      [[5, AchievementLevel.Gold]],
      intersection(NUTS, veggies).length,
    ),
    thirtyVeggies: achievementLevelHelper(
      [
        [40, AchievementLevel.Platinum],
        [30, AchievementLevel.Gold],
      ],
      veggies.length,
    ),
  }));

  const achievements = computed<Achievements>(() => ({
    challengeAccepted: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      completedChallenges.value,
    ),
    committed: achievementLevelHelper(
      [
        [52, AchievementLevel.Gold],
        [26, AchievementLevel.Silver],
        [12, AchievementLevel.Bronze],
      ],
      weeks.value.length,
    ),
    completionist: achievementLevelHelper(
      [
        [150, AchievementLevel.Gold],
        [80, AchievementLevel.Silver],
        [40, AchievementLevel.Bronze],
      ],
      uniqueVeggies.value.length,
    ),
    experimenterBean: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BEANS, uniqueVeggies.value).length,
    ),
    experimenterFruit: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(FRUITS, uniqueVeggies.value).length,
    ),
    experimenterGrain: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(GRAINS, uniqueVeggies.value).length,
    ),
    experimenterLeafy: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(LEAFIES, uniqueVeggies.value).length,
    ),
    experimenterMushroom: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(MUSHROOMS, uniqueVeggies.value).length,
    ),
    experimenterRoot: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(ROOTS, uniqueVeggies.value).length,
    ),
    experimenterVegetable: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(VEGETABLES, uniqueVeggies.value).length,
    ),
    hotStreak: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      hotStreak.value,
    ),
    thousandsEven:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 0
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    thousandsOdd:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 1
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    ...weeklyAchievements.value(),
  }));

  // Actions
  const toggleVeggie = (targetVeggie: string) =>
    toggleVeggieForWeek(targetVeggie, currentWeekStart.value);

  const toggleVeggieForWeek = (targetVeggie: string, weekStart: DateTime) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies: [targetVeggie],
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else if (!targetWeek.veggies.includes(targetVeggie)) {
      targetWeek.veggies = [...targetWeek.veggies, targetVeggie];
    } else {
      targetWeek.veggies = difference(targetWeek.veggies, [targetVeggie]);
    }
  };

  const setVeggiesForWeek = (veggies: string[], weekStart: DateTime = currentWeekStart.value) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies,
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else {
      targetWeek.veggies = veggies;
    }
  };

  const $reset = () => {
    startDate.value = null;
    weeks.value = [];
    challenges.value = [];
  };

  return {
    achievements,
    allVeggies,
    atMostVeggies,
    challenges,
    completedChallenges,
    currentChallenge,
    currentVeggies,
    favorites,
    getWeekStarts,
    hotStreak,
    over30Veggies,
    startDate,
    suggestions,
    uniqueVeggies,
    veggiesForWeek,
    weeklyAchievements,
    weeks,
    setVeggiesForWeek,
    toggleVeggie,
    toggleVeggieForWeek,
    $reset,
  };
});
