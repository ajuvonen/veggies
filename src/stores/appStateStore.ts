import {ref, watchEffect} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {useStorage} from '@vueuse/core';
import type {Settings} from '@/utils/types';
import {useAchievements} from '@/hooks/achievements';
import {useActivityStore} from '@/stores/activityStore';

type Message = {
  id: string;
  text: string;
};

export const useAppStateStore = defineStore('appState', () => {
  const {advance, achievements, reset} = useAchievements();
  const {uniqueVeggies} = storeToRefs(useActivityStore());

  watchEffect(() => {
    advance(uniqueVeggies.value, 0);
  });

  // State refs
  const settings = useStorage<Settings>(
    'veggies-settings',
    {
      locale: 'en',
    },
    localStorage,
    {
      mergeDefaults: true,
    },
  );

  const messages = ref<Message[]>([]);

  const addToastMessage = (text: string) => {
    messages.value.push({
      id: crypto.randomUUID(),
      text,
    });
  };

  const removeToastMessage = (targetId: string) => {
    messages.value = messages.value.filter(({id}) => id !== targetId);
  };

  const $reset = () => {
    messages.value = [];
    settings.value = {
      locale: 'en',
    };
    reset();
  };

  return {
    messages,
    settings,
    achievements,
    addToastMessage,
    removeToastMessage,
    $reset,
  };
});
