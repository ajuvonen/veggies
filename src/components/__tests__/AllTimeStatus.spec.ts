import {describe, it, expect, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import {DateTime} from 'luxon';
import {useActivityStore} from '@/stores/activityStore';
import AllTimeStatus from '@/components/AllTimeStatus.vue';

describe('AllTimeStatus', () => {
  let activityStore: ReturnType<typeof useActivityStore>;

  beforeEach(() => {
    activityStore = useActivityStore();
  });

  it('renders', () => {
    const wrapper = mount(AllTimeStatus);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows total weeks', () => {
    activityStore.startDate = DateTime.now().startOf('week').minus({weeks: 2});
    const wrapper = mount(AllTimeStatus);
    expect(wrapper.findByTestId('all-time-weeks').text()).toBe('In Total 3 Weeks');
  });

  it('shows unique veggies', () => {
    activityStore.startDate = DateTime.now().startOf('week').minus({weeks: 1});
    activityStore.weeks.push(
      {
        startDate: DateTime.now().startOf('week').minus({weeks: 1}),
        veggies: ['apple', 'cucumber', 'potato'],
      },
      {
        startDate: DateTime.now().startOf('week'),
        veggies: ['apple', 'cucumber', 'rice', 'fennel'],
      },
    );
    const wrapper = mount(AllTimeStatus);
    expect(wrapper.findByTestId('all-time-unique').text()).toBe('In Total 5 Unique Veggies');
  });

  it('shows weeks with over 30 veggies', () => {
    activityStore.startDate = DateTime.now().startOf('week').minus({weeks: 2});
    activityStore.weeks.push(
      {
        startDate: DateTime.now().startOf('week').minus({weeks: 2}),
        veggies: [...Array(31)],
      },
      {
        startDate: DateTime.now().startOf('week').minus({weeks: 1}),
        veggies: [...Array(29)],
      },
      {
        startDate: DateTime.now().startOf('week'),
        veggies: [...Array(30)],
      },
    );
    const wrapper = mount(AllTimeStatus);
    expect(wrapper.findByTestId('all-time-over-30').text()).toBe('Over 30 Veggies in 2 Weeks');
  });

  it('shows highest number of veggies', () => {
    activityStore.startDate = DateTime.now().startOf('week').minus({weeks: 2});
    activityStore.weeks.push(
      {
        startDate: DateTime.now().startOf('week').minus({weeks: 2}),
        veggies: [...Array(31)],
      },
      {
        startDate: DateTime.now().startOf('week').minus({weeks: 1}),
        veggies: [...Array(29)],
      },
      {
        startDate: DateTime.now().startOf('week'),
        veggies: [...Array(30)],
      },
    );
    const wrapper = mount(AllTimeStatus);
    expect(wrapper.findByTestId('all-time-at-most').text()).toBe('At Most 31 Veggies in a Week');
  });
});
