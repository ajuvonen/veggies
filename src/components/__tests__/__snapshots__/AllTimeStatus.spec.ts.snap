// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AllTimeStatus > renders 1`] = `
"<div data-v-900f68cf="" class="status__container">
  <div data-v-900f68cf="" class="status__item" data-test-id="all-time-weeks"><span data-v-900f68cf="">In Total</span>
    <div data-v-a0e2a650="" data-v-900f68cf="" class="statistic-share__container"><span data-v-a0e2a650="">1</span><button data-v-24e38f3d="" data-v-a0e2a650="" class="button button--text h-fit" aria-flowto="all-time-status-totalWeeks">
        <!--v-if--><svg data-v-a0e2a650="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
          <path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z" class="transition-all"></path>
        </svg>
      </button></div> <span data-v-900f68cf="" id="all-time-status-totalWeeks">Weeks</span>
  </div>
  <div data-v-900f68cf="" class="status__item" data-test-id="all-time-over-30"><span data-v-900f68cf="">Over 30 Veggies in</span>
    <div data-v-a0e2a650="" data-v-900f68cf="" class="statistic-share__container"><span data-v-a0e2a650="">0</span><button data-v-24e38f3d="" data-v-a0e2a650="" class="button button--text h-fit" aria-flowto="all-time-status-over30Veggies">
        <!--v-if--><svg data-v-a0e2a650="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
          <path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z" class="transition-all"></path>
        </svg>
      </button></div> <span data-v-900f68cf="" id="all-time-status-over30Veggies">Weeks</span>
  </div>
  <div data-v-900f68cf="" class="status__item" data-test-id="all-time-unique"><span data-v-900f68cf="">In Total</span>
    <div data-v-a0e2a650="" data-v-900f68cf="" class="statistic-share__container"><span data-v-a0e2a650="">0</span><button data-v-24e38f3d="" data-v-a0e2a650="" class="button button--text h-fit" aria-flowto="all-time-status-uniqueVeggies">
        <!--v-if--><svg data-v-a0e2a650="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
          <path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z" class="transition-all"></path>
        </svg>
      </button></div> <span data-v-900f68cf="" id="all-time-status-uniqueVeggies">Unique Veggies</span>
  </div>
  <div data-v-900f68cf="" class="status__item" data-test-id="all-time-at-most"><span data-v-900f68cf="">At Most</span>
    <div data-v-a0e2a650="" data-v-900f68cf="" class="statistic-share__container"><span data-v-a0e2a650="">0</span><button data-v-24e38f3d="" data-v-a0e2a650="" class="button button--text h-fit" aria-flowto="all-time-status-atMostVeggies">
        <!--v-if--><svg data-v-a0e2a650="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
          <path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z" class="transition-all"></path>
        </svg>
      </button></div> <span data-v-900f68cf="" id="all-time-status-atMostVeggies">Veggies in a Week</span>
  </div>
  <div data-v-900f68cf="" class="status__item col-2" data-test-id="all-time-challenges"><span data-v-900f68cf="">Completed</span>
    <div data-v-a0e2a650="" data-v-900f68cf="" class="statistic-share__container"><span data-v-a0e2a650="">0</span><button data-v-24e38f3d="" data-v-a0e2a650="" class="button button--text h-fit" aria-flowto="all-time-status-completedChallenges">
        <!--v-if--><svg data-v-a0e2a650="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
          <path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z" class="transition-all"></path>
        </svg>
      </button></div> <span data-v-900f68cf="" id="all-time-status-completedChallenges">Weekly Challenges</span>
  </div>
</div>"
`;
